### ISSUE TITLE
[PyTorch] torch.compile causes _default_get_amax to produce wrong results

### ISSUE THREAD

**Issue by Marks101** [NVIDIA/TransformerEngine#600]
Hello,
we noticed training instabilities in some of our FP8 trainings that use `transformer_engine.pytorch.Linear` to build a custom transformer stack. Looking into the checkpoints, we noticed that `scaling_bwd` deviates between tensor parallel ranks. We managed to build a minimal example that shows how the scaling factor starts to deviate:
<details>
  <summary>Script that shows how scaling_bwd deviates between tensor parallel ranks</summary>

```python
import os
import torch
import transformer_engine as te

world_size = int(os.environ["WORLD_SIZE"])
rank = int(os.environ["RANK"])

torch.cuda.set_device(rank)
torch.distributed.init_process_group(backend="nccl", world_size=world_size, rank=rank)
tp_group = torch.distributed.new_group(ranks=range(world_size))

fp8_recipe = te.common.recipe.DelayedScaling(
    margin=0,
    interval=1,
    fp8_format=te.common.recipe.Format.HYBRID,
    amax_history_len=1024,
    amax_compute_algo="max",
)

hidden_size = 2048
sequence_length = 2048

linear_1 = te.pytorch.Linear(hidden_size, hidden_size,
                            parallel_mode="column",
                            tp_group=tp_group)

linear_2 = te.pytorch.Linear(hidden_size, hidden_size,
                             parallel_mode="row",
                             tp_group=tp_group)

torch.manual_seed(32)
x = torch.randn(sequence_length, hidden_size).cuda()

for i in range(100):
    with te.pytorch.fp8_autocast(fp8_recipe=fp8_recipe, fp8_group=tp_group):
        output = linear_2(linear_1(x))

    output_grad = torch.randn_like(output)
    output.backward(output_grad)

    print(f"rank={rank}-i={i}: linear_1 scaling_bwd={linear_1.fp8_meta['scaling_bwd'].scale[0].item()}, "
          f"linear_2 scaling_bwd={linear_2.fp8_meta['scaling_bwd'].scale[0].item()}")
```
</details>

When running this script (torchrun --nproc-per-node 2 script.py) we see deviations for `scaling_bwd` on the second linear:
```
rank=1-i=99: linear_1 scaling_bwd=8756.66796875, linear_2 scaling_bwd=8896.1728515625
rank=0-i=99: linear_1 scaling_bwd=8756.66796875, linear_2 scaling_bwd=10947.583984375
```

Digging deeper we figured that this issue might arise from `torch.compile` which is applied inside `amax_and_scale_update()`. Specifically `roll()` and `fill_()` inside `_update_amax_history()` seem to be problematic. Here is another minimal example that shows the issue:

```python
import torch
import transformer_engine as te

#amax_history = torch.ones(1024, 3, device="cuda")
#amax_history, _ = te.pytorch.fp8._default_get_amax(amax_history, "max")

amax_history = torch.ones(1024, 2, device="cuda")
amax_history, _ = te.pytorch.fp8._default_get_amax(amax_history, "max")
print(amax_history)
assert torch.all(amax_history[0, :] == 0.0)
assert torch.all(amax_history[1:, :] == 1.0)
```

The script has to be executed multiple times. On the first pass the assert usually passes. If `torch.jit.script` is used instead of `torch.compile` (`NVTE_TORCH_COMPILE=0`) everything seems to be fine. Looking at the triton code (`TORCH_COMPILE_DEBUG=1`) my assumption is that the mask is missing during load and store, but I did not validate this assumption. This would make it an issue with PyTorch. But still I would kindly ask you to try to reproduce my findings first before reporting this. Overall I am very puzzled how this became a problem for our current training and was fine in many other scenarios. But it seems that `torch.compile` produces a generalized kernel if it approaches different column sizes of `amax_history`. If you uncomment the two lines in the example above everything is fine as well.

As a quick fix I found that replacing `amax_history[0].fill_(0.0)` with `amax_history = torch.cat([torch.zeros_like(amax_history[:1]), amax_history[1:]])` works but prodces an additional kernel launch.

I ran everyhing on a H100 DGX inside the official NVCR PyTorch Container (nvcr.io/nvidia/pytorch:23.12-py3).

**Comment by ksivaman** [comment]
Thanks for the detailed comment. I haven't been able to repro it on my end so far, will investigate! @Marks101

**Comment by ptrendx** [comment]
@ptrblck FYI, we may need some help from your side after Kirthi's initial investigation.

**Comment by Marks101** [comment]
Thank you for the quick replies!

I have been digging deeper and reduced the issue to this piece of code (now independent of transformer-engine):
```python
import torch

def roll_and_zero_first_row(x):
    x = torch.roll(x, -1, 0)
    x[0].fill_(0.0)
    return x

inp = torch.ones(1024, 2, device="cuda")
compiled = torch.compile(roll_and_zero_first_row)

out_eager = roll_and_zero_first_row(inp)
out_compiled = compiled(inp)

print("compiled", out_compiled)
print("eager", out_eager)
assert torch.all(out_eager == out_compiled)
```

I run the script as follows:
```bash
docker run --gpus all --rm -v /tmp/test.py:/tmp/test.py nvcr.io/nvidia/pytorch:23.12-py3 python /tmp/test.py
```

The assertion fails non-deterministically. This issue is not reproducable in each run.

One thing I noticed after setting `TORCH_COMPILE_DEBUG=1` is that the problem seems to occur if XBLOCK=256 is the fastest kernel in the triton heuristics. XBLOCK=128 seems to run properly. This is the log output I am refering to:
```
[2024-01-15 09:49:27,109] torch._inductor.triton_heuristics: [DEBUG] XBLOCK: 256, num_warps: 4, num_ctas: 1, num_stages: 1, enable_warp_specialization: False, enable_persistent: False: 0.005248, nreg 12, nspill 0, #shared-mem 0
[2024-01-15 09:49:27,109] torch._inductor.triton_heuristics: [DEBUG] XBLOCK: 128, num_warps: 4, num_ctas: 1, num_stages: 1, enable_warp_specialization: False, enable_persistent: False: 0.005344, nreg 10, nspill 0, #shared-mem 0
```

It would be really great if you could try to reproduce these findings. Thank you!

**Comment by ksivaman** [comment]
I'm able to reproduce this as well. @ptrblck Could someone from the PyTorch team take a look?

**Comment by Marks101** [comment]
By now I am pretty sure that this issue might arise from triton. I opend an issue there https://github.com/openai/triton/issues/2941

**Comment by ksivaman** [comment]
@Marks101 I'm observing a reliably reproducible failure associated specifically with the input size of 1024. The generated triton kernel omits the mask for checking bounds during the `triton.language.load` instruction resulting in the observed difference in amaxes. This is not the case for any other inputs size I've looked at thus far

**Comment by Marks101** [comment]
@ksivaman yes, this is in accordance with my findings. It is specifically with the input shape (1024, 2) that the mask is omitted. But from my understanding it should be fine to omit it because the the element count of 2048 is evenly divisible by the block sizes 128 or 256.

**Comment by ksivaman** [comment]
I haven't done a fully exhaustive search but (1024, 2), (2048, 1), (2048, 2), (4096, 1), (4096, 2), (8192, 1), (8192, 2) are some of the other sizes that fail.

**Comment by Marks101** [comment]
@ksivaman thanks for bringing up this PR. This looks like a decent fix that should do it for the moment
